Add Role `v7`

tigrato opened the pull request:
This PR introduces a new role version - `v7` - to support other
Kubernetes resources such as `Deployments`, `Statefulsets`... while
keeping compatibility with previous role versions.

This PR adds the following validations:
- `v6`, `v5`, `v4`: `kubernetes_resources` can only include `kind: pod`
  resources
- `v7`: `kubernetes_resources` can include wildcards or other
  resources (to be introduced by https://github.com/gravitational/teleport/pull/26244).

It also adds a role version downgrade (v13 clients connecting to v14
auth server) where we downgrade the role's version and remove unsupported
kubernetes_resources from the resource list.

Part of https://github.com/gravitational/teleport/issues/20811

espadolini reviewed (dismissed):
`buf breaking` with `WIRE` (or `WIRE_JSON`) settings:

```
% buf breaking . --against '.git#branch=origin/master'
api/proto/teleport/legacy/client/proto/authservice.proto:69:5:Field "9" on message "Event" changed type from "types.RoleV6" to "types.RoleV7".
api/proto/teleport/legacy/client/proto/authservice.proto:216:12:Field "6" on message "OpenSSHCertRequest" changed type from "types.RoleV6" to "types.RoleV7".
api/proto/teleport/legacy/client/proto/authservice.proto:1055:12:Field "1" on message "GetRolesResponse" changed type from "types.RoleV6" to "types.RoleV7".
api/proto/teleport/legacy/client/proto/authservice.proto:2437:66:RPC "GetCurrentUserRoles" on service "AuthService" changed response type from "types.RoleV6" to "types.RoleV7".
api/proto/teleport/legacy/client/proto/authservice.proto:2571:40:RPC "GetRole" on service "AuthService" changed response type from "types.RoleV6" to "types.RoleV7".
api/proto/teleport/legacy/client/proto/authservice.proto:2575:18:RPC "UpsertRole" on service "AuthService" changed request type from "types.RoleV6" to "types.RoleV7".
```

tigrato committed:
Introduce version `v7` to Role resources

This PR introduces a new role version - `v7` - to support other
Kubernetes resources such as `Deployments`, `Statefulsets`... while
keeping compatibility with previous role versions.

This PR adds the following validations:
- `v6`, `v5`, `v4`: `kubernetes_resources` can only include `kind: pod`
  resources
- `v7`: `kubernetes_resources` can include wildcards or other
  resourceces (to be introduced by #26244.

It also adds a role version downgrade (v13 clients connecting to v14
auth server) where we downgrade the role's version and remove unsuported
kubernetes_resources from the resource list.

Part of #20811

jakule commented on api/types/role_test.go:
```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2020 Gravitational, Inc.
```
```suggestion
Copyright 2023 Gravitational, Inc.
```

jakule commented on integrations/lib/testing/integration/bootstrap.go:
```diff
@@ -45,7 +45,8 @@ func (bootstrap *Bootstrap) AddUserWithRoles(name string, roles ...string) (type
 
 func (bootstrap *Bootstrap) AddRole(name string, spec types.RoleSpecV6) (types.Role, error) {
-	role, err := types.NewRole(name, spec)
+	// TODO: Remove this once Test Integration is updated to build tctl.
```
Can you create an issue for that, assign it to the owner, and link it here? Otherwise, this will never be addressed.

tigrato replied:
done in [0911cb0](https://github.com/gravitational/teleport/pull/26583/commits/0911cb035617ca17e4f96e0cfecbcb3aac82c8f0)

jakule reviewed (commented):
@tigrato We got into a habit of incrementing the role version quite recently. Can you create (if we already don't have it) a place in our docs to describe the differences between each version and when they were introduced? I assume that users are/will be confused with a new version with almost every release.
Waiting with the approval until @espadolini comment is addressed.

espadolini commented:
Work is underway to rename `RoleV6` (which is actually v3 to v6 and - with this PR - v7) into something with a more sensible name that's stable across the supported versions of Teleport, as part of #26688.

nklaassen commented on lib/auth/grpcserver.go:
```diff
@@ -1915,22 +1925,59 @@ func (g *GRPCServer) DeleteAllKubernetesServers(ctx context.Context, req *proto.
+		}
+	}
+
+	if clientVersion == nil || clientVersion.LessThan(*MinSupportedRoleV7Version) {
+		log.Debugf(`Client version "%s" is unknown or less than 14.0.0, converting role to v6`, clientVersionString)
```
I don't think we should downgrade if the client version is unknown. Teleport clients have been reporting their versions for long enough now that if they don't, they probably don't support v6 roles anyway. This is more likely to be a third-party client and downgrading the role in that case will just be confusing I think

tigrato commented:
@espadolini @jakule @nklaassen 

I rebased this PR on top of Nic's #27244

Can you please re-review this?

tigrato commented on api/types/role.go:
```diff
@@ -390,11 +396,46 @@ func (r *RoleV6) SetKubeGroups(rct RoleConditionType, groups []string) {
+			return []KubernetesResource{{Kind: Wildcard, Name: Wildcard, Namespace: Wildcard}}
+		default:
+			for _, resource := range KubernetesResourcesKinds {
+				// Ignore Pod resources for older roles since by default they always have access to pods.
```
Yeah, I had in mind the role v5

tigrato commented on api/types/role.go:
```diff
@@ -390,11 +396,46 @@ func (r *RoleV6) SetKubeGroups(rct RoleConditionType, groups []string) {
+		return resources
+	default:
+		switch {
+		// If role does not have kube labels, return empty list.
+		case len(r.Spec.Allow.KubernetesLabels) == 0 && len(r.Spec.Allow.KubernetesLabelsExpression) == 0:
+			return nil
```
Yes, this is just to reduce the number of elements we return.
The role has no matcher for the cluster so it won't be used.

tigrato commented on api/types/role.go:
```diff
@@ -948,12 +989,27 @@ func (r *RoleV6) CheckAndSetDefaults() error {
+		}
+	case V7:
+		// Kubernetes resources default to {kind:*, name:*, namespace:*} for v7 roles.
+		if len(r.Spec.Allow.KubernetesResources) == 0 && len(r.Spec.Allow.KubernetesLabels) > 0 {
```
Yeah, that's much better.
Introduced `HasLabelMatchers`

```go
// HasLabelMatchers returns true if the role has label matchers for the
// specified resource kind and condition (allow/deny).
// If the kind is not supported, false is returned.
func (r *RoleV6) HasLabelMatchers(rct RoleConditionType, kind string) bool {
	lm, err := r.GetLabelMatchers(rct, kind)
	return err == nil && !lm.Empty()
}
```

tigrato commented on lib/auth/grpcserver.go:
```diff
@@ -1991,26 +1996,125 @@ func maybeDowngradeRoleLabelExpressions(ctx context.Context, role *types.RoleV6,
+				types.Allow,
+				types.Labels{},
+			)
+			// Clear out the allow list so that the V6 role doesn't include unknown
+			// resources in the allow list.
+			downgraded.SetKubeResources(types.Allow, nil)
```
Unfortunately, it's not possible because the client requesting this downgrade isn't able to filter any other resources so it will allow all `kubectl get service` without any proper filtering or control. Since we remove the matcher, the role won't be used and so we can just exclude the `kubernetes_resources`.

The only way of preventing users to access the other resources is by denying access to the cluster using this role.

tigrato commented on lib/auth/grpcserver.go:
```diff
@@ -1991,26 +1996,125 @@ func maybeDowngradeRoleLabelExpressions(ctx context.Context, role *types.RoleV6,
+		if downgraded.Metadata.Labels == nil {
+			downgraded.Metadata.Labels = make(map[string]string, 1)
+		}
+		downgraded.Metadata.Labels[types.TeleportDowngradedLabel] = reason
```
I tried to avoid appending the messages because they will be gigantic.

Do you feel it's important?

nklaassen replied:
nah not important

jakule commented:
@tigrato UTs are still failing.

jakule commented on api/types/role.go:
```diff
@@ -390,11 +396,52 @@ func (r *RoleV6) SetKubeGroups(rct RoleConditionType, groups []string) {
+// to other resources. This is a simple optimization to reduce the number of resources.
+// Finally, if the older role version is not a wildcard, then it returns the pod resources as is
+// and append the other supported resources - KubernetesResourcesKinds - for Role v7.
+func (r *RoleV6) convertKubernetesResourcesBetweenRoleVersions(resources []KubernetesResource) []KubernetesResource {
```
Can you add this v13/14 downgrade logic to the test plan? I think we should test it right before the release once again.

jakule commented on api/types/role.go:
```diff
@@ -390,11 +396,52 @@ func (r *RoleV6) SetKubeGroups(rct RoleConditionType, groups []string) {
+	switch r.Version {
+	case V7:
+		return resources
+	case V6, V5, V4, V3:
```
Can you comment on why v1 and v2 are not listed?

jakule commented on api/types/role_test.go:
```diff
@@ -143,3 +143,198 @@ func TestAccessReviewConditionsIsEmpty(t *testing.T) {
+		want                []KubernetesResource
+		assertErrorCreation require.ErrorAssertionFunc
+	}{
+		// TODO(tigrato): add more tests once we support other kubernetes resources.
```
Isn't this always an implicit TODO? Test more stuff when we add it :sweat_smile:

jakule commented on integrations/lib/tctl/resources.go:
```diff
@@ -89,7 +89,7 @@ func (res *streamResource) UnmarshalJSON(raw []byte) error {
 	case types.KindRole:
 		switch header.Version {
-		case types.V4, types.V5, types.V6:
+		case types.V4, types.V5, types.V6, types.V7:
```
Why is v6 returned for v7 case?

jakule reviewed (commented):
A few minor comments. Overall looks good.
